parsel/selector.py

```diff
@@ -479,9 +474,12 @@ def re(
         Passing ``replace_entities`` as ``False`` switches off these
         replacements.
         """
-        return extract_regex(
-            regex, self.get(), replace_entities=replace_entities
-        )
+        data = self.get()
+        if not isinstance(data, str):
+            data = json.dumps(
+                data, default="" if self.type == "json" else repr
+            )
```

Comment:
I am not convinced by this approach. I cannot think of a scenario where this is better than either adjusting the JMESPath expression to return strings (there is [to_string()](https://jmespath.org/proposals/functions.html#to-string)), or getting the data _as is_ and using something other than a regular expression to check its content.

I think it may be best to raise a `ValueError` if the input is not of a valid type for `re()`, just as `re.search` would do.

Reply:
I meant that _users_ can use `to_string` in the JMESPath queries if they wish, not that we should do that on our end automatically. I think we should raise `ValueError` on non-string input, as `re.search` would do.

Reply:
I don't have a strong opinion (as I didn't study the matter) but not converting things implicitly sounds better to me

Reply:
I don't have a strong opinion here. 

So, the question is what to do when .jmespath() returns non-string (e.g. a number, or a dictionary), and then .re is called? Raising ValueError seems reasonable, but running re on some stringified version of the data is also fine with me (parsel Selectors almost never fail currently). 

+0.5 for ValueError though :)

Let's say developer wrote some code which uses .re to extract data from a string field in JSON. Then, JSON is changed, so that the field is no longer a string - e.g. it's a dictionary, or a list now. In this case it might be better to fail with ValueError, instead of silently starting to convert data to string, because the conversion is likely to produce incorrect output.